File: modules/xq-domain-test-mcp/src/xq_mcp/tools/runtime_config.py
from mcp.server.fastmcp import FastMCP

from xq_mcp.runtime import RuntimeState


def configure_environment(
    state: RuntimeState,
    *,
    environment: str,
    api_base_url: str,
    api_token: str | None = None,
) -> dict[str, object]:
    return state.configure(
        environment=environment,
        api_base_url=api_base_url,
        api_token=api_token,
    )


def get_environment(state: RuntimeState) -> dict[str, object]:
    return state.status()


def clear_environment(state: RuntimeState) -> dict[str, object]:
    return state.clear()


class RuntimeConfigTools:
    _state: RuntimeState

    def __init__(self, state: RuntimeState) -> None:
        self._state = state

    def register(self, mcp: FastMCP) -> None:
        state = self._state

        @mcp.tool(name="configure_environment")
        def configure_environment_tool(
            environment: str,
            api_base_url: str,
            api_token: str | None = None,
        ) -> dict[str, object]:
            """Configure runtime environment parameters for this MCP process."""
            return configure_environment(
                state,
                environment=environment,
                api_base_url=api_base_url,
                api_token=api_token,
            )

        @mcp.tool(name="get_environment")
        def get_environment_tool() -> dict[str, object]:
            """Read redacted runtime environment status."""
            return get_environment(state)

        @mcp.tool(name="clear_environment")
        def clear_environment_tool() -> dict[str, object]:
            """Clear runtime environment parameters from memory."""
            return clear_environment(state)

        _ = (configure_environment_tool, get_environment_tool, clear_environment_tool)

File: modules/xq-domain-test-mcp/src/tools/runtime-config.ts
import { z } from "zod/v4";

import type { RuntimeEnvironmentStatus, RuntimeState } from "../runtime/runtime-state.js";
import type { McpTool } from "./mcp-tool.js";

export type ConfigureEnvironmentInput = {
  environment: string;
  api_base_url: string;
  api_token?: string;
};

export type ConfigureEnvironmentOutput = {
  status: "configured";
  configured: true;
  environment: string;
  api_base_url: string;
  has_api_token: boolean;
};

export type GetEnvironmentInput = Record<string, never>;
export type GetEnvironmentOutput = RuntimeEnvironmentStatus;
export type ClearEnvironmentInput = Record<string, never>;
export type ClearEnvironmentOutput = {
  status: "cleared";
  was_configured: boolean;
  configured: false;
};

const environmentStatusSchema = z.object({
  configured: z.boolean(),
  environment: z.string().optional(),
  api_base_url: z.string().optional(),
  has_api_token: z.boolean().optional()
});

const configuredStatusSchema = z.object({
  configured: z.literal(true),
  environment: z.string(),
  api_base_url: z.string(),
  has_api_token: z.boolean()
});

export class ConfigureEnvironmentTool
  implements McpTool<ConfigureEnvironmentInput, ConfigureEnvironmentOutput>
{
  readonly name = "configure_environment";
  readonly category = "runtime_config";
  readonly title = "Configure Environment";
  readonly description = "Configure runtime environment parameters for this MCP process.";
  readonly inputSchema = z.object({
    environment: z.string().min(1),
    api_base_url: z.string().min(1),
    api_token: z.string().optional()
  }) satisfies z.ZodType<ConfigureEnvironmentInput>;
  readonly outputSchema = configuredStatusSchema.extend({
    status: z.literal("configured")
  }) satisfies z.ZodType<ConfigureEnvironmentOutput>;

  constructor(private readonly runtime: RuntimeState) {}

  execute(input: ConfigureEnvironmentInput): ConfigureEnvironmentOutput {
    return this.runtime.configure(input);
  }
}

export class GetEnvironmentTool implements McpTool<GetEnvironmentInput, GetEnvironmentOutput> {
  readonly name = "get_environment";
  readonly category = "runtime_config";
  readonly title = "Get Environment";
  readonly description = "Read redacted runtime environment status.";
  readonly annotations = { readOnlyHint: true };
  readonly inputSchema = z.object({}) satisfies z.ZodType<GetEnvironmentInput>;
  readonly outputSchema = environmentStatusSchema satisfies z.ZodType<GetEnvironmentOutput>;

  constructor(private readonly runtime: RuntimeState) {}

  execute(_input: GetEnvironmentInput): GetEnvironmentOutput {
    return this.runtime.status();
  }
}

export class ClearEnvironmentTool implements McpTool<ClearEnvironmentInput, ClearEnvironmentOutput> {
  readonly name = "clear_environment";
  readonly category = "runtime_config";
  readonly title = "Clear Environment";
  readonly description = "Clear runtime environment parameters from memory.";
  readonly inputSchema = z.object({}) satisfies z.ZodType<ClearEnvironmentInput>;
  readonly outputSchema = z.object({
    status: z.literal("cleared"),
    was_configured: z.boolean(),
    configured: z.literal(false)
  }) satisfies z.ZodType<ClearEnvironmentOutput>;

  constructor(private readonly runtime: RuntimeState) {}

  execute(_input: ClearEnvironmentInput): ClearEnvironmentOutput {
    return this.runtime.clear();
  }
}
